akka-http/src/main/scala/akka/http/javadsl/server/directives/CodingDirectives.scala

```diff
@@ -5,42 +5,84 @@
-import akka.http.scaladsl.server._
-
-import scala.annotation.varargs
 import akka.http.scaladsl
-import akka.http.impl.server.RouteStructure
+import akka.http.javadsl.model.headers.HttpEncoding
+import java.util.function.Supplier
+import akka.http.scaladsl.server.{ Directives ⇒ D }
+import akka.http.javadsl.server.JavaScalaTypeEquivalence._
+import scala.collection.JavaConverters._
 
 abstract class CodingDirectives extends CacheConditionDirectives {
   /**
-   * Wraps the inner routes with encoding support. The response will be encoded
-   * using one of the predefined coders, `Gzip`, `Deflate`, or `NoCoding` depending on
-   * a potential [[akka.http.javadsl.model.headers.AcceptEncoding]] header from the client.
+   * Rejects the request with an UnacceptedResponseEncodingRejection
+   * if the given response encoding is not accepted by the client.
    */
-  @varargs def encodeResponse(innerRoute: Route, moreInnerRoutes: Route*): Route =
-    RouteStructure.EncodeResponse(CodingDirectives._DefaultCodersToEncodeResponse)(innerRoute, moreInnerRoutes.toList)
+  def responseEncodingAccepted(encoding: HttpEncoding, inner: Supplier[Route]): Route = ScalaRoute {
+    D.responseEncodingAccepted(encoding) {
+      inner.get.toScala
+    }
+  }
 
   /**
-   * A directive that Wraps its inner routes with encoding support.
-   * The response will be encoded using one of the given coders with the precedence given
-   * by the order of the coders in this call.
+   * Encodes the response with the encoding that is requested by the client via the `Accept-
+   * Encoding` header. The response encoding is determined by the rules specified in
+   * http://tools.ietf.org/html/rfc7231#section-5.3.4.
+   *
+   * If the `Accept-Encoding` header is missing or empty or specifies an encoding other than
+   * identity, gzip or deflate then no encoding is used.
+   */
+  def encodeResponse(inner: Supplier[Route]): Route = ScalaRoute {
+    D.encodeResponse {
+      inner.get.toScala
+    }
+  }
+
+  /**
+   * Encodes the response with the encoding that is requested by the client via the `Accept-
+   * Encoding` header. The response encoding is determined by the rules specified in
+   * http://tools.ietf.org/html/rfc7231#section-5.3.4.
+   *
+   * If the `Accept-Encoding` header is missing then the response is encoded using the `first`
+   * encoder.
+   *
+   * If the `Accept-Encoding` header is empty and `NoCoding` is part of the encoders then no
+   * response encoding is used. Otherwise the request is rejected.
    *
-   * In any case, a potential [[akka.http.javadsl.model.headers.AcceptEncoding]] header from the client
-   * will be respected (or otherwise, if no matching .
+   * If [encoders] is empty, no encoding is performed.
    */
-  @varargs def encodeResponse(coders: Coder*): Directive =
-    Directives.custom(RouteStructure.EncodeResponse(coders.toList))
+  // TODO provide Java API for creating Encoder, by extracting a Java interface from the Scala trait
```

Comment:
Looks like there is quite a bit of stuff in there. Since it's not possible to create a custom `Encoder` with the current DSL either, could we postpone this to a future PR?